### ISSUE TITLE
flutter_test does not use the latest version of test_api

### ISSUE THREAD

**Issue by iapicca** [flutter/flutter#81294]
### to reproduce
- create a package that depends on the latest version of [test](https://pub.dev/packages/test) (at the time of writing `test: ^1.17.2`)
- create an app based on the latest master (see `doctor` below)
- add the package to the `dev_dependencies`
- run `flutter pub get`
### expected result

- `flutter_test` should be `up-to-date` with the latest version of `test`
- a possible error message should be "understandable"

### actual result

- flutter_test is **_not_** `up-to-date` with `test` more specifically [test_api](https://pub.dev/packages/test_api) (`^0.4.0` at the time of writing) 
- the error message is **_not informative_** and **_misleading_** 
    -  `Minimum version ("0.4.0") ` minimum version of what?
    -  immediately direct the user to filing an `issue`

<details>
<summary>flutter pub get</summary>

```console
$ flutter pub get
Invalid argument(s): Minimum version ("0.4.0") must be less than maximum ("0.4.0-0").
package:pub_semver/src/version_range.dart 67:7      new VersionRange    
package:pub/src/solver/reformat_ranges.dart 56:27   _reformatTerm       
package:pub/src/solver/reformat_ranges.dart 34:26   reformatRanges.<fn> 
dart:_internal                                      ListIterable.toList 
package:pub/src/solver/reformat_ranges.dart 35:12   reformatRanges      
package:pub/src/solver/reformat_ranges.dart 133:25  _reformatCause      
package:pub/src/solver/reformat_ranges.dart 36:7    reformatRanges      
package:pub/src/solver/reformat_ranges.dart 133:25  _reformatCause      
package:pub/src/solver/reformat_ranges.dart 36:7    reformatRanges      
package:pub/src/solver/reformat_ranges.dart 133:25  _reformatCause      
package:pub/src/solver/reformat_ranges.dart 36:7    reformatRanges      
package:pub/src/solver/reformat_ranges.dart 133:25  _reformatCause      
package:pub/src/solver/reformat_ranges.dart 36:7    reformatRanges      
package:pub/src/solver/reformat_ranges.dart 133:25  _reformatCause      
package:pub/src/solver/reformat_ranges.dart 36:7    reformatRanges      
package:pub/src/solver/version_solver.dart 308:24   VersionSolver._resolveConflict
package:pub/src/solver/version_solver.dart 129:27   VersionSolver._propagate
package:pub/src/solver/version_solver.dart 93:11    VersionSolver.solve.<fn>
===== asynchronous gap ===========================                      
package:pub/src/source/hosted.dart 635:14           BoundHostedSource.withPrefetching.<fn>
===== asynchronous gap ===========================                      
package:pub/src/rate_limited_scheduler.dart 110:14  RateLimitedScheduler.withPrescheduling
===== asynchronous gap ===========================                      
package:pub/src/source/hosted.dart 634:12           BoundHostedSource.withPrefetching
===== asynchronous gap ===========================                      
package:pub/src/solver/version_solver.dart 90:14    VersionSolver.solve 
===== asynchronous gap ===========================                      
package:pub/src/entrypoint.dart 233:18              Entrypoint.acquireDependencies
This is an unexpected error. Please run                                 
                                                                        
    dart pub --trace get --no-precompile                                
                                                                        
and include the logs in an issue on https://github.com/dart-lang/pub/issues/new
Running "flutter pub get" in <****some_app****> ...                              
pub get failed (1; and include the logs in an issue on https://github.com/dart-lang/pub/issues/new)
```

</details>

only `flutter pub get -v` refer the actual package/packages causing the issue
<details>
<summary>flutter pub get -v (partial)</summary>

```console
[   +1 ms] SLVR:   fact: no versions of <****some_package****> match >=0.4.0 <0.4.0+0
[        ] SLVR:   conflict: no versions of <****some_package****> match >=0.4.0 <0.4.0+0
[        ] SLVR:   ! <****some_package****> >=0.4.0 <0.4.0+0 is partially satisfied by not <****some_package****> >=0.4.0+0
[        ] SLVR:   ! which is caused by "<****some_package****> >=0.4.0+0 depends on test ^1.17.1"
[        ] SLVR:   ! thus: <****some_package****> >=0.4.0 requires test ^1.17.1
[        ] SLVR:   ! not test ^1.17.1 is satisfied by not test >=1.17.0
[        ] SLVR:   ! which is caused by "test >=1.17.0 depends on test_api 0.4.0"
[        ] SLVR:   ! thus: <****some_package****> >=0.4.0 requires test_api 0.4.0
[        ] SLVR:   ! not test_api 0.4.0 is satisfied by test_api 0.3.0
[        ] SLVR:   ! which is caused by "every version of flutter_test from sdk depends on test_api 0.3.0"
[        ] SLVR:   ! thus: <****some_package****> >=0.4.0 is incompatible with flutter_test from sdk
[        ] SLVR:   ! flutter_test any from sdk is satisfied by flutter_test any from sdk
[        ] SLVR:   ! which is caused by "<****some_app****> depends on flutter_test any from sdk"
[        ] SLVR:   ! thus: <****some_package****> >=0.4.0 is forbidden
[        ] SLVR:   ! <****some_package****> >=0.4.0 is satisfied by <****some_package****> ^0.4.0
[        ] SLVR:   ! which is caused by "<****some_app****> depends on <****some_package****> ^0.4.0"
[        ] SLVR:   ! thus: version solving failed
[   +2 ms] SLVR: Version solving took 0:00:02.930735 seconds.
```

</details>
<details>
<summary>doctor</summary>

```console
[✓] Flutter (Channel master, 2.3.0-1.0.pre.105, on Linux, locale en_US.UTF-8)
    • Flutter version 2.3.0-1.0.pre.105 at /home/yakforward/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3eb5736134 (2 hours ago), 2021-04-27 07:49:08 -0400
    • Engine revision 9e4b4e23a3
    • Dart version 2.14.0 (build 2.14.0-47.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /home/yakforward/Android/Sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /snap/android-studio/101/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 12.0.0-1ubuntu1
    • cmake version 3.18.4
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 4.1)
    • Android Studio at /snap/android-studio/101/android-studio
    • Flutter plugin version 55.1.1
    • Dart plugin version 201.9335
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Linux
    • Chrome (web)    • chrome • web-javascript • Google Chrome 90.0.4430.85

• No issues found!

```

</details>

### proposal

- would be nice to have some automation that updates `flutter_test` when a new release of `test_api` is released 
- the error message could be more clear

**Comment by darshankawar** [comment]
@iapicca Hi,

Can you please provide an example for `create a package that depends on the latest version of test` ?
Thanks.

**Comment by iapicca** [comment]
@darshankawar 
here are more details to reproduce the issue :

- create a package and an app

```console
$ mkdir issue_81294
$ cd issue_81294
$ flutter create example
$ flutter create --template=package test_package
```

- change the `pubspec` of `test_package` to 
```yaml
name: test_package
version: 0.0.1

environment:
  sdk: ">=2.12.0 <3.0.0"

dependencies:
  test: ^1.17.2
```
- to avoid tree-shaking change the content of `lib/test_package.dart` to
```dart
library test_package;

import 'package:test/test.dart';

void main() {
  test('empty', () {});
}
```

- change the pubspec of `example` to 
```yaml
name: example
version: 1.0.0+1

environment:
  sdk: ">=2.12.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter

dev_dependencies:
  flutter_test:
    sdk: flutter
  test_package: any

dependency_overrides:
  test_package:
    path: ../test_package

flutter:
  uses-material-design: true
```

- run `flutter pub get`

#

here is the folder, just run `flutter create .` 
[issue_81294.zip](https://github.com/flutter/flutter/files/6389942/issue_81294.zip)

**Comment by darshankawar** [comment]
Thanks @iapicca for the detailed steps to replicate. I get below error upon running `flutter pub get`:

```
dhs@Dhss-MacBook-Pro example % flutter pub get
Because every version of test_package from path depends on test ^1.17.2 which depends on test_api 0.4.0, every version of test_package from path requires test_api 0.4.0.
And because every version of flutter_test from sdk depends on test_api 0.3.0, test_package from path is incompatible with flutter_test from sdk.
So, because example depends on both flutter_test any from sdk and test_package from path, version solving failed.
Running "flutter pub get" in example...                                 
pub get failed (1; So, because example depends on both flutter_test any from sdk
and test_package from path, version solving failed.)
dhs@Dhss-MacBook-Pro example % 
```

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.3.0-1.0.pre.143, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.3.0-1.0.pre.143 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision aecd5e0302 (6 hours ago), 2021-04-28 00:53:57 -0300
    • Engine revision 92ef7b59c8
    • Dart version 2.14.0 (build 2.14.0-48.0.dev)

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • Mac OS X 10.15.4 19E2269
      darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 90.0.4430.85

! Doctor found issues in 1 category.
```

</details>

**Comment by iapicca** [comment]
@darshankawar 
thanks mate, one more thing:
I don't know why but now the error message is totally different from the one I got [before](https://github.com/flutter/flutter/issues/81294#issue-868887027)
as it is now is totally clear in my opinion the error message label is unnecessary

**Comment by natebosch** [comment]
cc @jonahwilliams - updating to the latest `test_api` will require updating the `mockito` dependency in `flutter_tool`.

I think there are long term plans to remove use of mockito? In the mean time I think the breaking change in version `5.0.0` is unlikely to impact the usage in flutter_tool so it should be safe to upgrade.

**Comment by jonahwilliams** [comment]
The long term plans are short term plans now, once CI is working again and we land a few more PRs we're down to ~18 or usages.  Since we're long past a branch point for stable, we'd rather keep plugging away since the newer mockito versions introduce many more additional dependencies anyway.

**Comment by mit-mit** [comment]
Can we tag this with the 'May milestone'? I'd like for this to have some soak time before the next stable.
